src/v/storage/disk_log_impl.cc

```diff
@@ -259,8 +259,16 @@ ss::future<> disk_log_impl::do_compact(compaction_config cfg) {
       });
     if (segit != _segs.end()) {
         auto seg = *segit;
-        return storage::internal::self_compact_segment(seg, cfg, _probe)
-          .finally([seg] { seg->mark_as_finished_self_compaction(); });
+        auto f = ss::now();
+        if (_stm_manager) {
+            f = _stm_manager->ensure_snapshot_exists(
```

Comment:
this snapshot occurs before a segment is compacted. if all the segments have been compacted, then below we try and compact pairs of segments, but it looks like we aren't taking after those compaction events. is that intentional?

Reply:
yes, the snapshotting has two goals:
 
  - to avoid reading the whole log on bootstrap
  - to save information and to GC the log before the point of snapshotting

Right now we have two forms of GC: log segments eviction (log_eviction_stm) and compaction.

As I understand compaction happens in phases. During the first phase we compact segments and then we do cross-segment compaction of already compacted segments. Since disk_log_impl calls ensure_snapshot_exists before each segment compaction then when things come to cross-segment compaction there is no need to take a snapshot because the snapshot of the last segment already covers the affected offset range.